include/exadg/incompressible_navier_stokes/time_integration/time_int_bdf_consistent_splitting.cpp

```diff
@@ -0,0 +1,895 @@
+
+  /*
+   *  Right-hand side viscous term:
+   *  If there is no nonlinearity, we solve a linear system of equations, where
+   *  inhomogeneous parts of boundary face integrals of the viscous operator
+   *  have to be shifted to the right-hand side of the equation.
+   */
+  if(this->param.viscous_problem() and not(this->param.nonlinear_problem_has_to_be_solved()))
+  {
+    pde_operator->rhs_add_viscous_term(rhs, this->get_next_time());
+  }
+}
+
+template<int dim, typename Number>
+void
+TimeIntBDFConsistentSplitting<dim, Number>::penalty_step()
+{
+  if(this->param.use_divergence_penalty == true or this->param.use_continuity_penalty == true)
+  {
+    dealii::Timer timer;
+    timer.restart();
+
+    // compute right-hand-side vector
+    VectorType rhs(velocity_np);
+    pde_operator->apply_mass_operator(rhs, velocity_np);
+
+    // extrapolate velocity to time t_n+1 and use this velocity field to
+    // calculate the penalty parameter for the divergence and continuity penalty term
+    VectorType velocity_extrapolated(velocity_np);
+    velocity_extrapolated = 0.0;
+    for(unsigned int i = 0; i < velocity.size(); ++i)
+      velocity_extrapolated.add(this->extra.get_beta(i), velocity[i]);
+
+    pde_operator->update_projection_operator(velocity_extrapolated, this->get_time_step_size());
+
+    // right-hand side term: add inhomogeneous contributions of continuity penalty operator to
+    // rhs-vector if desired
+    if(this->param.use_continuity_penalty and this->param.continuity_penalty_use_boundary_data)
+      pde_operator->rhs_add_projection_operator(rhs, this->get_next_time());
+
+    // solve linear system of equations
+    bool const update_preconditioner =
+      this->param.update_preconditioner_projection and
+      ((this->time_step_number - 1) %
+         this->param.update_preconditioner_projection_every_time_steps ==
+       0);
+
+    if(this->use_extrapolation == false)
+      velocity_np = 0;
+
+    unsigned int const n_iter =
+      pde_operator->solve_projection(velocity_np, rhs, update_preconditioner);
+
+    iterations_penalty.first += 1;
+    iterations_penalty.second += n_iter;
+
+    // write output
+    if(this->print_solver_info() and not(this->is_test))
+    {
+      this->pcout << std::endl << "Solve penalty step:";
+      print_solver_info_linear(this->pcout, n_iter, timer.wall_time());
+    }
+
+    this->timer_tree->insert({"Timeloop", "Penalty step"}, timer.wall_time());
+  }
+}
+
+template<int dim, typename Number>
+void
+TimeIntBDFConsistentSplitting<dim, Number>::prepare_vectors_for_next_timestep()
+{
+  Base::prepare_vectors_for_next_timestep();
+
+  push_back(velocity);
+  velocity[0].swap(velocity_np);
+
+  push_back(pressure);
+  pressure[0].swap(pressure_np);
+
+  // We also have to care about the history of velocity Dirichlet boundary conditions.
+  // Note that velocity_dbc_np has already been updated.
+  push_back(velocity_dbc);
+  velocity_dbc[0].swap(velocity_dbc_np);
```

Comment:
We should eventually (i.e., not in this PR) get rid of that global vector, and have a specialized vector for quadrature points at the Dirichlet boundaries only. It should be reasonably simple to set up, but having a brittle mechanism for the boundary dofs into the cell. Think about the corner where multiple faces are at the Dirichlet boundary; this works for most cases but if you use discontinuous data leads to trouble.

Reply:
We had those in some cases of FSI problems (or more generally, with moving domains), where this gets injected from another field. If I remember correctly, we had different versions while developing the various algorithms for FSI, and some of the infrastructure in ExaDG was written before the remote point evaluation of deal.II (and related functions) were created, so there are some selections that could be streamlined here I believe. So yes, I agree we should work towards proving all information we need at quadrature points, and only have that option.